Documentation/specs/corehost.md

```diff
@@ -0,0 +1,117 @@
+# corehost runtime/assembly resolution
+
+The shared host locates assemblies and native libraries using a combination of: Servicing Index, Files in the application folder (aka "app-local") and files from package caches.
+
+## Definitions and Formats
+
+### Terms/Notes
+
+* The term "Library" (Title Case) is used throughout this document to refer to a NuGet Package. We use this term because it is used in the code to represent multiple things: Packages, Projects and Framework Assemblies are all types of "Library".
+* All of the assembly resolution here refers to setting up the default assembly load context for the runtime. Further dynamic loads (plugins, etc.) can still have custom resolution logic provided by using an `AssemblyLoadContext` in the managed code. Essentially, we are setting up the necessary assemblies to launch the `Program.Main` (and all the assemblies that are statically-referenced by that).
+
+### Servicing Index
+
+The servicing index is loaded when the `DOTNET_SERVICING` environment variable is non-empty. When this variable is non-empty, it points to a directory that will be the **Servicing Root**. There may be platform-specific default locations, to be determined later.
+
+An index file is located at the path defined by `$DOTNET_SERVICING/servicing_index.txt`. In this file are a series of lines of one of the following formats:
+
+```
+# Lines starting with a '#' and blank lines are ignored.
+
+# Identifies an asset from a NuGet Package that has been serviced.
+package|[Package ID]|[Package Version]|[Original Asset Relative Path]=[New Asset Path, relative to Servicing Root]
```

Comment:
Would a better format be a directory structure?
package/package_id/version/{relative.txt, redirect.txt}

This would make the look-ups O(1) for each assembly file in the TPA.